src/scales/scale.time.js

```diff
@@ -361,27 +361,55 @@ function generate(min, max, capacity, options) {
  * Returns the right and left offsets from edges in the form of {left, right}.
  * Offsets are added when the `offset` option is true.
  */
-function computeOffsets(table, ticks, min, max, options) {
+function computeOffsets(table, ticks, data, min, max, options) {
+	var pos = [];
+	var minInterval = 1;
+	var timeOpts = options.time;
+	var barThickness = options.barThickness;
 	var left = 0;
 	var right = 0;
-	var upper, lower;
-
-	if (options.offset && ticks.length) {
-		if (!options.time.min) {
-			upper = ticks.length > 1 ? ticks[1] : max;
-			lower = ticks[0];
-			left = (
-				interpolate(table, 'time', upper, 'pos') -
-				interpolate(table, 'time', lower, 'pos')
-			) / 2;
-		}
-		if (!options.time.max) {
-			upper = ticks[ticks.length - 1];
-			lower = ticks.length > 1 ? ticks[ticks.length - 2] : min;
-			right = (
-				interpolate(table, 'time', upper, 'pos') -
-				interpolate(table, 'time', lower, 'pos')
-			) / 2;
+	var i, ilen, curr, prev, length, width;
+
+	if (options.offset) {
+		data.forEach(function(timestamp) {
+			if (timestamp >= min && timestamp <= max) {
+				pos.push(interpolate(table, 'time', timestamp, 'pos'));
+			}
+		});
+
+		length = pos.length;
+		if (length) {
+			// Calculate minInterval
+			if (!barThickness) {
+				[data, ticks].forEach(function(timestamps) {
+					for (i = 0, ilen = timestamps.length; i < ilen; ++i) {
+						curr = interpolate(table, 'time', timestamps[i], 'pos');
+						minInterval = i > 0 ? Math.min(minInterval, curr - prev) : minInterval;
+						prev = curr;
+					}
+				});
+			}
+
+			if (!timeOpts.min) {
+				if (length === 1) {
+					width = (1 - pos[0]) * 2;
+				} else if (barThickness) {
+					width = pos[1] - pos[0];
```

Comment:
What it there's only a single data point? I think `pos[1]` will be undefined in that case

Reply:
Nevermind. I see you already handled that